File: word_frequency.rb
# word_frequency.rb
require 'set'

class WordFrequency
  def initialize
    @stop_words_en = Set.new(%w[
      a an the and or but for nor on at to by in with without of per via plus minus up down off over under
    ])
    @stop_words_ru = Set.new(%w[
      и в во не что он на я с со как а то все она так его но да ты к у же вы за бы по только ее мне было
      вот от меня еще нет о из ему теперь когда даже ну вдруг ли если уже или ни быть был него до вас нибудь
      опять уж вам ведь там потом себя ничего ей может они тут где есть надо ней для мы тебя их чем была сам
      чтоб без будто чего раз тоже себе под будет ж тогда кто этот того потому этого какой совсем ним здесь
      этом один почти мой тем чтобы нее сейчас были куда зачем всех можно при наконец сегодня любой два об
      другой хоть после над больше тот через эти нас про всего них какая много разве три эту моя впрочем
      хорошо свою этой перед иногда лучше чуть том нельзя такой более всё конечно всю между ваше начал свое
      свои ваши твою тобой тобою твоя твои твое вами нами
    ])
    @use_stop_words = true
    @top_n = 10
    @language = 'en'
  end

  attr_accessor :top_n, :use_stop_words, :language

  def stop_words
    @language == 'ru' ? @stop_words_ru : @stop_words_en
  end

  def process_text(text)
    words = text.downcase.scan(/\b\w+\b/)
    total = words.size
    filtered = @use_stop_words ? words.reject { |w| stop_words.include?(w) } : words
    freq = Hash.new(0)
    filtered.each { |w| freq[w] += 1 }
    sorted = freq.sort_by { |w, c| [-c, w] }
    top = sorted.first(@top_n).map { |w, c| { word: w, count: c } }
    { top: top, total: total, unique: freq.size }
  end

  def display(result)
    if result[:top].empty?
      puts "No words found after filtering."
      return
    end
    puts "\nTop #{result[:top].size} words (total: #{result[:total]}, unique: #{result[:unique]}):"
    result[:top].each do |item|
      pct = item[:count].to_f / result[:total] * 100
      puts "  #{item[:word]}: #{item[:count]} (#{'%.1f' % pct}%)"
    end
  end
end

def main
  analyzer = WordFrequency.new
  puts "=== Word Frequency Finder ==="
  loop do
    puts "\n1. Enter text manually"
    puts "2. Load from file"
    puts "3. Set number of top words (current: #{analyzer.top_n})"
    puts "4. Toggle stop words (currently: #{analyzer.use_stop_words ? 'on' : 'off'})"
    puts "5. Set language (current: #{analyzer.language})"
    puts "6. Exit"
    print "Choose: "
    choice = gets.chomp.strip
    case choice
    when '1'
      puts "Enter your text (end with empty line):"
      lines = []
      loop do
        line = gets.chomp
        break if line.empty?
        lines << line
      end
      text = lines.join("\n")
      unless text.empty?
        result = analyzer.process_text(text)
        analyzer.display(result)
      end
    when '2'
      print "Enter file path: "
      fname = gets.chomp.strip
      unless File.exist?(fname)
        puts "File not found."
        next
      end
      content = File.read(fname)
      result = analyzer.process_text(content)
      analyzer.display(result)
    when '3'
      print "Enter number of top words: "
      n = gets.chomp.to_i
      if n <= 0
        puts "Must be positive."
      else
        analyzer.top_n = n
      end
    when '4'
      analyzer.use_stop_words = !analyzer.use_stop_words
      puts "Stop words toggled."
    when '5'
      print "Enter language (en/ru): "
      lang = gets.chomp.strip.downcase
      if ['en', 'ru'].include?(lang)
        analyzer.language = lang
        puts "Language set to #{lang}."
      else
        puts "Invalid language."
      end
    when '6'
      puts "Goodbye!"
      break
    else
      puts "Invalid choice."
    end
  end
end

main if __FILE__ == $0

File: word_frequency.swift
// word_frequency.swift
import Foundation

class WordFrequency {
    private let stopWordsEn: Set<String> = [
        "a", "an", "the", "and", "or", "but", "for", "nor", "on", "at",
        "to", "by", "in", "with", "without", "of", "per", "via", "plus",
        "minus", "up", "down", "off", "over", "under"
    ]
    private let stopWordsRu: Set<String> = [
        "и", "в", "во", "не", "что", "он", "на", "я", "с", "со", "как",
        "а", "то", "все", "она", "так", "его", "но", "да", "ты", "к",
        "у", "же", "вы", "за", "бы", "по", "только", "ее", "мне", "было",
        "вот", "от", "меня", "еще", "нет", "о", "из", "ему", "теперь",
        "когда", "даже", "ну", "вдруг", "ли", "если", "уже", "или", "ни",
        "быть", "был", "него", "до", "вас", "нибудь", "опять", "уж", "вам",
        "ведь", "там", "потом", "себя", "ничего", "ей", "может", "они",
        "тут", "где", "есть", "надо", "ней", "для", "мы", "тебя", "их",
        "чем", "была", "сам", "чтоб", "без", "будто", "чего", "раз",
        "тоже", "себе", "под", "будет", "ж", "тогда", "кто", "этот",
        "того", "потому", "этого", "какой", "совсем", "ним", "здесь",
        "этом", "один", "почти", "мой", "тем", "чтобы", "нее", "сейчас",
        "были", "куда", "зачем", "всех", "можно", "при", "наконец",
        "сегодня", "любой", "два", "об", "другой", "хоть", "после",
        "над", "больше", "тот", "через", "эти", "нас", "про", "всего",
        "них", "какая", "много", "разве", "три", "эту", "моя", "впрочем",
        "хорошо", "свою", "этой", "перед", "иногда", "лучше", "чуть",
        "том", "нельзя", "такой", "более", "всё", "конечно", "всю",
        "между", "ваше", "начал", "свое", "свои", "ваши", "твою",
        "тобой", "тобою", "твоя", "твои", "твое", "вами", "нами"
    ]
    var useStopWords = true
    var topN = 10
    var language = "en"

    private var stopWords: Set<String> {
        return language == "ru" ? stopWordsRu : stopWordsEn
    }

    func processText(_ text: String) -> (top: [(word: String, count: Int)], total: Int, unique: Int) {
        let words = text.lowercased().components(separatedBy: CharacterSet.letters.inverted)
            .filter { !$0.isEmpty }
        let total = words.count
        let filtered = useStopWords ? words.filter { !stopWords.contains($0) } : words
        var freq: [String: Int] = [:]
        for w in filtered {
            freq[w, default: 0] += 1
        }
        let sorted = freq.sorted { (a, b) -> Bool in
            if a.value == b.value { return a.key < b.key }
            return a.value > b.value
        }
        let top = sorted.prefix(topN).map { (word: $0.key, count: $0.value) }
        return (top, total, freq.count)
    }

    func display(result: (top: [(word: String, count: Int)], total: Int, unique: Int)) {
        if result.top.isEmpty {
            print("No words found after filtering.")
            return
        }
        print("\nTop \(result.top.count) words (total: \(result.total), unique: \(result.unique)):")
        for item in result.top {
            let pct = Double(item.count) / Double(result.total) * 100
            print("  \(item.word): \(item.count) (\(String(format: "%.1f", pct))%)")
        }
    }
}

func main() {
    let analyzer = WordFrequency()
    print("=== Word Frequency Finder ===")
    while true {
        print("\n1. Enter text manually")
        print("2. Load from file")
        print("3. Set number of top words (current: \(analyzer.topN))")
        print("4. Toggle stop words (currently: \(analyzer.useStopWords ? "on" : "off"))")
        print("5. Set language (current: \(analyzer.language))")
        print("6. Exit")
        print("Choose: ", terminator: "")
        guard let choice = readLine()?.trimmingCharacters(in: .whitespaces) else { continue }
        switch choice {
        case "1":
            print("Enter your text (end with empty line):")
            var lines: [String] = []
            while true {
                guard let line = readLine() else { break }
                if line.isEmpty { break }
                lines.append(line)
            }
            let text = lines.joined(separator: "\n")
            if !text.isEmpty {
                let result = analyzer.processText(text)
                analyzer.display(result: result)
            }
        case "2":
            print("Enter file path: ", terminator: "")
            guard let fname = readLine()?.trimmingCharacters(in: .whitespaces) else { break }
            let fileURL = URL(fileURLWithPath: fname)
            guard let content = try? String(contentsOf: fileURL, encoding: .utf8) else {
                print("File not found or unreadable.")
                break
            }
            let result = analyzer.processText(content)
            analyzer.display(result: result)
        case "3":
            print("Enter number of top words: ", terminator: "")
            if let nStr = readLine(), let n = Int(nStr), n > 0 {
                analyzer.topN = n
            } else {
                print("Must be a positive integer.")
            }
        case "4":
            analyzer.useStopWords.toggle()
            print("Stop words toggled.")
        case "5":
            print("Enter language (en/ru): ", terminator: "")
            if let lang = readLine()?.trimmingCharacters(in: .whitespaces).lowercased(),
               lang == "en" || lang == "ru" {
                analyzer.language = lang
                print("Language set to \(lang).")
            } else {
                print("Invalid language.")
            }
        case "6":
            print("Goodbye!")
            return
        default:
            print("Invalid choice.")
        }
    }
}

main()
